### PR TITLE
Minimize IRep Serialization memory consumption

### PR DESCRIPTION
**Pull request by celinval**
### Description of changes: 

Compiling the std-library was taking 11+G of memory. We root caused it to the codegen serialization code that was generating the entire json string in memory before dumping to a file.

In this change, I am using serde to serialize the irep with minimum buffering. This reduces the memory consumption to 4.5G while compiling the STD library. This can still be optimized by either improving the irep memory consumption or getting rid of the irep completely.

### Resolved issues:

This solves #622

### Testing:

* How is this change tested? Everywhere

* Is this a refactor change? Yes

### Checklist
- [ ] Each commit message has a non-empty body, explaining why the change was made
- [ ] Methods or procedures are documented
- [ ] Regression or unit tests are included, or existing tests cover the modified code
- [ ] My PR is restricted to a single feature or bugfix

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

File: compiler/rustc_codegen_rmc/src/compiler_interface.rs
Comment (tedinski) [inline compiler/rustc_codegen_rmc/src/compiler_interface.rs:164]: Wrap out_file with a BufWriter and see if that cures your performance problems.

Comment (adpaco-aws) [inline compiler/rustc_codegen_rmc/src/compiler_interface.rs:164]: I was experimenting with `serde` by doing:
```
            let out_file = ::std::fs::File::create(&symtab_filename).unwrap();
            let writer = BufWriter::new(out_file);
            serde_json::to_writer(writer, &result.symtab);
```
The main differences are:
 1. Using a BufWriter to handle writing.
 2. Serializing the symbol table directly instead of the irep representation of the single table.

This has brought memory usage down to 1GB if I have not made any mistakes in measuring.

Not sure if this is related to (2), but conceptually I think it is wrong to transform the symbol table with `to_irep` before writing the JSON file. The serialization of a symbol table is the irep representation itself. Of course, this solution would need writing custom serializers for some objects in the symbol table, but I think it is worth it.

Comment (celinval) [reply compiler/rustc_codegen_rmc/src/compiler_interface.rs:164]: That's a good idea. Let me try.

Comment (celinval) [reply compiler/rustc_codegen_rmc/src/compiler_interface.rs:164]: For sure, I agree that this is the right path. I used irep because it was the quick and dirty way to unblock me and a fairly simple change. No harm done.

File: compiler/cbmc/src/irep/serialize.rs
Comment (danielsn) [inline compiler/cbmc/src/irep/serialize.rs:93]: Would it make sense to use one of the `Symbol` constructors?

File: compiler/rustc_codegen_rmc/src/compiler_interface.rs
Comment (danielsn) [inline compiler/rustc_codegen_rmc/src/compiler_interface.rs:172]: We do this bit more than once. Should it be a helper function?

File: scripts/std-lib-regression.sh
Comment (danielsn) [inline scripts/std-lib-regression.sh:9]: What auto-formatter are you using, and can we get it enabled in CI so we don't have whitespace wars?

File: compiler/cbmc/src/irep/serialize.rs
Comment (danielsn) [inline compiler/cbmc/src/irep/serialize.rs:95]: Should we add something here to make sure that it recurses correctly?

File: compiler/rustc_codegen_rmc/src/compiler_interface.rs
Comment (danielsn) [reply compiler/rustc_codegen_rmc/src/compiler_interface.rs:172]: Would also help make sure that temporaries were dropped as soon as the helper function exited.

File: scripts/std-lib-regression.sh
Comment (danielsn) [inline scripts/std-lib-regression.sh:2]: For git hygine, it would be better to put this as its own PR.

File: compiler/cbmc/src/irep/serialize.rs
Comment (celinval) [reply compiler/cbmc/src/irep/serialize.rs:93]: sure.

File: compiler/rustc_codegen_rmc/src/compiler_interface.rs
Comment (celinval) [reply compiler/rustc_codegen_rmc/src/compiler_interface.rs:172]: ok

File: scripts/std-lib-regression.sh
Comment (celinval) [reply scripts/std-lib-regression.sh:9]: For bash scripts I just use `vim`. :)

File: scripts/std-lib-regression.sh
Comment (celinval) [reply scripts/std-lib-regression.sh:2]: Yeah, it's already in it's own PR (#593), but that PR was blocked by the github memory restriction. I could remove this commit from here, push this fix first and get back to the other PR later.

File: compiler/cbmc/src/irep/serialize.rs
Comment (celinval) [reply compiler/cbmc/src/irep/serialize.rs:93]: I think you might be talking about the goto_program::Symbol, but this code is for the irep::Symbol which doesn't have any constructor.

**Review by danielsn — APPROVED** [review]

**Review by adpaco-aws — APPROVED** [review]

**Pull request merged**
